libbeat/beat/event_test.go

```diff
@@ -18,351 +18,569 @@
+				require.NoError(t, err)
+				require.Equal(t, tc.val, actual)
+				has, err := event.HasKey(tc.key)
+				require.NoError(t, err)
+				require.True(t, has)
+			})
+		}
+
+		t.Run("type conflict", func(t *testing.T) {
+			event := &Event{
+				Meta: mapstr.M{
+					"a": 9,
+					"c": 10,
+				},
+				Fields: mapstr.M{
+					"a": 9,
+					"c": 10,
+				},
+			}
+
+			_, err := event.PutValue("a.c", 10)
+			require.Error(t, err)
+			require.Equal(t, "expected map but type is int", err.Error())
+			_, err = event.PutValue("a.value", 9)
+			require.Error(t, err)
+			require.Equal(t, "expected map but type is int", err.Error())
+		})
 
-	b.Run("get sub-key", func(b *testing.B) {
-		evt := newEmptyEvent()
-		evt.Meta = newMeta()
+		t.Run("hierarchy", func(t *testing.T) {
+			event := &Event{
+				Fields: mapstr.M{
+					"a.b": 1,
+				},
+			}
+			err := event.Delete("a.b")
+			require.NoError(t, err)
+
+			prev, err := event.PutValue("a.b.c", 1)
+			require.NoError(t, err)
+			require.Nil(t, prev)
+
+			expFields := mapstr.M{
+				"a": mapstr.M{
+					"b": mapstr.M{
+						"c": 1,
+					},
+				},
+			}
 
-		v, err := evt.GetValue("@metadata._id")
+			require.Equal(t, expFields, event.Fields)
+		})
 
-		assert.NoError(b, err)
-		assert.Equal(b, id, v)
+		t.Run("SetID", func(t *testing.T) {
+			event := &Event{}
+			event.SetID("unique")
+			require.Equal(t, "unique", event.Meta["_id"])
+		})
 	})
 
-	b.Run("delete", func(b *testing.B) {
-		evt := newEmptyEvent()
-		evt.Meta = newMeta()
-
-		err := evt.Delete("@metadata")
-
-		assert.NoError(b, err)
-		assert.Nil(b, evt.Meta)
+	t.Run("SetErrorWithOption", func(t *testing.T) {
+		cloned := event.Clone()
+		cloned.SetErrorWithOption("message", false, "data", "field")
+		require.Equal(t, event, cloned)
+		expEvent := cloned.Clone()
+		expEvent.Fields[ErrorFieldKey] = mapstr.M{
+			"message": "message",
+			"field":   "field",
+			"data":    "data",
+			"type":    "json",
+		}
+		cloned.SetErrorWithOption("message", true, "data", "field")
+		require.Equal(t, expEvent, cloned)
 	})
 
-	b.Run("delete sub-key", func(b *testing.B) {
-		evt := newEmptyEvent()
-		evt.Meta = newMeta()
-
-		err := evt.Delete("@metadata._id")
-
-		assert.NoError(b, err)
-		assert.Empty(b, evt.Meta)
-	})
+	t.Run("DeepUpdate", func(t *testing.T) {
+		newTs := time.Now().Add(time.Hour)
+		update := map[string]interface{}{
+			TimestampFieldKey: newTs,
+			MetadataFieldKey: map[string]interface{}{
+				"metaLevel0Map": mapstr.M{ // mix types on purpose, should support both
+					"metaLevel1Map": map[string]interface{}{
+						"new1": "newmetavalue1",
+					},
+				},
+				"metaLevel0Value": "metareplaced1",
+				"new2":            "newmetavalue2",
+			},
+			"fieldsLevel0Map": map[string]interface{}{
+				"fieldsLevel1Map": mapstr.M{
+					"new3": "newfieldsvalue1",
+				},
+				"newmap": map[string]interface{}{
+					"new4": "newfieldsvalue2",
+				},
+			},
+			"fieldsLevel0Value": "fieldsreplaced1",
+		}
+
```

Comment:
If you're mixing types on purpose, what about also using `map[string]any`?

Reply:
`any` is just an alias for `interface{}`, is not it? what difference would it make?